lenskit-slim/src/main/java/org/lenskit/slim/LinearRegressionHelper.java

```diff
@@ -0,0 +1,143 @@
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.slim;
+
+import it.unimi.dsi.fastutil.longs.*;
+import org.lenskit.util.math.Scalars;
+import java.util.Map;
+import static java.lang.Math.abs;
+
+
+/**
+ * Several simple vector algebra used by SLIM learning process
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+public final class LinearRegressionHelper {
+    /**
+     * soft-Thresholding for coordinate descent
+     * @param z
+     * @param gamma
+     * @return
+     */
+    public static double softThresholding(double z, double gamma) {
+        if (z > 0 && gamma < abs(z)) return z - gamma;
+        if (z < 0 && gamma < abs(z)) return z + gamma; // comment this line to get non-negative weights during learning process
+        return 0.0;
+    }
+
+    /**
+     * Transpose of matrix
+     * Convert a row view of rating matrix to a column view and vice versa
+     * After conversion, the key of empty row/column in the input matrix will be eliminated in the transpose matrix.
+     * So calling transposeMap(transposeMap(X)).keySet() may not equal to X.keySet() unless X has no mapping of key to empty Long2Double map.
+     *
+     * @param matrix
+     * @return
+     */
+    public static Long2ObjectMap<Long2DoubleMap> transposeMap(Long2ObjectMap<Long2DoubleMap> matrix) {
+        Long2ObjectMap<Long2DoubleMap> mapT = new Long2ObjectOpenHashMap<>();
+
+        for (Map.Entry<Long, Long2DoubleMap> rowEntry : matrix.entrySet()) {
+            long rowNum = rowEntry.getKey();
+            Long2DoubleMap rowValue = rowEntry.getValue();
+            for (Map.Entry<Long, Double> entry : rowValue.entrySet()) {
+                long colNum = entry.getKey();
+                double Value = entry.getValue();
+                Long2DoubleMap rowOfMapT = mapT.get(colNum);
+                if (rowOfMapT == null) rowOfMapT = new Long2DoubleOpenHashMap();
+                rowOfMapT.put(rowNum, Value);
+                mapT.put(colNum, rowOfMapT);
+            }
+        }
+        return mapT;
+    }
+
+    /**
+     * add two vectors
+     * @param a
+     * @param b
+     * @return
+     */
+    public static Long2DoubleMap addVectors(Long2DoubleMap a, Long2DoubleMap b) {
+        //Long2DoubleMap sumOfTwoVectors = Vectors.combine(a, b, 1.0, 0.0);
+        Long2DoubleMap sumOfTwoVectors = new Long2DoubleOpenHashMap(a);
+        for (Map.Entry<Long, Double> e : b.entrySet()) {
+            long key = e.getKey();
+            double value = e.getValue();
+            double sum = sumOfTwoVectors.get(key) + value;
+            sumOfTwoVectors.put(key, sum);
+        }
+        return sumOfTwoVectors;
+
+    }
+
+    /**
+     * element-wise multiplication of two vectors.
+     * @param a
+     * @param b
+     * @return
+     */
+    public static Long2DoubleMap multiply(Long2DoubleMap a, Long2DoubleMap b) {
```

Comment:
Maybe call it `ebeMultiply` to match the `commons-math` method name? But also, it should live in `Vectors`, even if it is only used in tests.